subsys/logging/log_backend_cavs_hda.c

```diff
@@ -0,0 +1,263 @@
+LOG_OUTPUT_DEFINE(log_output_cavs_hda, hda_log_out, log_buf, LOG_BUF_SIZE);
+
+static void hda_log_periodic(struct k_timer *tm)
+{
+	ARG_UNUSED(tm);
+
+	hda_log_flush();
+}
+
+static inline void dropped(const struct log_backend *const backend,
+			   uint32_t cnt)
+{
+	ARG_UNUSED(backend);
+
+	log_output_dropped_process(&log_output_cavs_hda, cnt);
+}
+
+static void panic(struct log_backend const *const backend)
+{
+	ARG_UNUSED(backend);
+
+	log_backend_std_panic(&log_output_cavs_hda);
+
+	panic_mode = true;
+}
+
+static int format_set(const struct log_backend *const backend, uint32_t log_type)
+{
+	ARG_UNUSED(backend);
+
+	log_format_current = log_type;
+	return 0;
+}
+
+static uint32_t format_flags(void)
+{
+	uint32_t flags = LOG_OUTPUT_FLAG_LEVEL | LOG_OUTPUT_FLAG_TIMESTAMP;
+
+	if (IS_ENABLED(CONFIG_LOG_BACKEND_FORMAT_TIMESTAMP)) {
+		flags |= LOG_OUTPUT_FLAG_FORMAT_TIMESTAMP;
+	}
+	return flags;
+}
+
+static volatile uint32_t counter;
+
+static void process(const struct log_backend *const backend,
+		union log_msg2_generic *msg)
+{
+	ARG_UNUSED(backend);
+
+	log_format_func_t log_output_func = log_format_func_t_get(log_format_current);
+
+	log_output_func(&log_output_cavs_hda, &msg->log, format_flags());
+}
+
+/**
+ * Lazily initialized, while the DMA may not be setup we continue
+ * to buffer log messages untilt he buffer is full.
+ */
+static void init(const struct log_backend *const backend)
+{
+	ARG_UNUSED(backend);
+
+	hda_log_buffered = 0;
+}
+
+const struct log_backend_api log_backend_cavs_hda_api = {
+	.process = process,
+	.dropped = IS_ENABLED(CONFIG_LOG_MODE_IMMEDIATE) ? NULL : dropped,
+	.panic = panic,
+	.format_set = format_set,
+	.init = init,
+};
+
+LOG_BACKEND_DEFINE(log_backend_cavs_hda, log_backend_cavs_hda_api, true);
+
+void cavs_hda_log_init(cavs_hda_log_hook_t fn, uint32_t channel)
+{
+	hook = fn;
+
+	int res;
+
+	hda_log_dev = device_get_binding("HDA_HOST_IN");
+	__ASSERT(hda_log_dev, "No valid DMA device found");
+
+	hda_log_chan = dma_request_channel(hda_log_dev, &channel);
+	__ASSERT(hda_log_chan >= 0, "No valid DMA channel");
+	__ASSERT(hda_log_chan == channel, "Not requested channel");
+
+	hda_log_buffered = 0;
+
+	/* configure channel */
+	struct dma_block_config hda_log_dma_blk_cfg = {
+		.block_size = CONFIG_LOG_BACKEND_CAVS_HDA_SIZE,
+		.source_address = (uint32_t)(uintptr_t)&hda_log_buf,
+	};
+
+	struct dma_config hda_log_dma_cfg = {
+		.channel_direction = MEMORY_TO_HOST,
+		.block_count = 1,
+		.head_block = &hda_log_dma_blk_cfg,
+		.source_data_size = 4,
+	};
+
+	res = dma_config(hda_log_dev, hda_log_chan, &hda_log_dma_cfg);
+	__ASSERT(res == 0, "DMA config failed");
+
+	res = dma_start(hda_log_dev, hda_log_chan);
+	__ASSERT(res == 0, "DMA start failed");
+
+	k_timer_init(&hda_log_timer, hda_log_periodic, NULL);
+	k_timer_start(&hda_log_timer,
+		      K_MSEC(CONFIG_LOG_BACKEND_CAVS_HDA_FLUSH_TIME),
+		      K_MSEC(CONFIG_LOG_BACKEND_CAVS_HDA_FLUSH_TIME));
+
+	initialized = true;
```

Comment:
Pedantic: the k_timer_start() call is scheduling a function to run on an uninitialized object here.  You want to flag it initialized before you call API functions on it, or even schedule API functions.  (Obviously in practice this won't matter)

Reply:
This is a fair point, initialized is meant to signify the DMA channel is available to be flushed to. dma_ready might be a better variable name for this, and moving it before the timer makes sense